java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java

```diff
@@ -2292,10 +2292,12 @@ private void readDictionaryStream(InStream in) throws IOException {
           int dictionaryBufferSize = dictionaryOffsets[dictionaryOffsets.length - 1];
           dictionaryBuffer = new byte[dictionaryBufferSize];
           int pos = 0;
-          int chunkSize = in.available();
-          byte[] chunkBytes = new byte[chunkSize];
+          //check if dictionary size is smaller than available stream size
+          // to avoid ArrayIndexOutOfBoundsException
+          int readSize = Math.min(in.available(), dictionaryBufferSize);
+          byte[] chunkBytes = new byte[readSize];
           while (pos < dictionaryBufferSize) {
-            int currentLength = in.read(chunkBytes, 0, chunkSize);
+            int currentLength = in.read(chunkBytes, 0, readSize);
             System.arraycopy(chunkBytes, 0, dictionaryBuffer, pos, currentLength);
             pos += currentLength;
```

Comment:
Would this fix be better?
```suggestion
            currentLength = Math.min(currentLength, dictionaryBufferSize - pos);
            System.arraycopy(chunkBytes, 0, dictionaryBuffer, pos, currentLength);
            pos += currentLength;
```

Reply:
> Actually this won't be a problem, because if the readSize is bigger than the available input, it just returns the available here.

The example I gave was specific to the LLAP environment. Could you test it out? I'm not worried about other cases, as available is always accurate.

Reply:
@guiyanakuang ok, I understand it now. I combined my changes with your suggestions. Combining this two reduces the memory usage also. With that this cases are covered:

Case 1: 
in.available() = 4096
dictionaryBuffer = 900
readSize = 900
It  creates a chunkBytes array only with 900 bytes (instead of 4096 bytes) and reads 900 bytes in one run.

Case2:
in.available() = 4k
dictionaryBufferSize = 5k
readSize = 4k
It  creates a chunkBytes with 4kB and reads 4kB at first run, then it reads 1kB and copies the remaining 1kB to the dictionaryBuffer.

Reply:
@guiyanakuang can we close this thread? Can you merge this PR?